# **Creating promises**

We create a promise when a certain task's completion time is uncertain or too long. For example, a network request.

A new promise is created by using the **Promise** constructor:

```js
const promise = new Promise((resolve, reject) => {
  if (Math.random() * 100 <= 90) {
    resolve("I am resolved!");
  }

  reject(new Error("10% of the times, I will fail"));
});
```

The constructor accepts a function with two parameters. This function is called an `executor` function and it describes
the computation being done. The parameters conventionally named `resolve` and `reject`, mark successful and unsuccessful
eventual completion of the executor function.

The `resolve` and `reject` are functions and are used to send back values to the promise object. When the computation
is successful or the future value is ready, we send the value back using `resolve`. We can say that the promise has been
resolved.

If it fails or encounters an error, we signal that by passing the error object in the `reject` function. We can say
that the promise has been rejected. `reject` accepts any value, however it's recommended to pass an `Error` object,
since it helps in debugging by viewing the stacktrace.

